### PR TITLE
[Extensions] Suggesting which extension to Load/Install

### PR DESCRIPTION
**Pull request by pdet**
This PR includes:
1) A CSV file with all currently supported extensions and where to find them (This is also used for the out-of-tree deployment)
2) A script that generates a map of all functions and their respective extensions
3) A CI that checks if the list is up to date
4) The necessary changes to suggest which extensions to install and load if the user tries to use one of their registered functions.

From this, is "trivial" (famous last words) to do automatic load/install, if we decide to do that at some point.

### CODE REVIEW

File: src/catalog/catalog.cpp [hunk 1]
```diff
@@ -243,7 +243,14 @@ CatalogException Catalog::CreateMissingEntryException(ClientContext &context, co
 		}
 	});
 	auto unseen_entry = SimilarEntryInSchemas(context, entry_name, type, unseen_schemas);
-
+	if (context.extensions_functions.find(entry_name) != context.extensions_functions.end()) {
+		auto extension_name = context.extensions_functions.find(entry_name)->second;
+		string exception_message = "Function with name " + entry_name +
+		                           " is not on the catalog, but it exists on the " + extension_name + "  extension!\n";
```
Comment (Mause) [inline src/catalog/catalog.cpp:249]:
```suggestion
		                           " is not on the catalog, but it exists in the " + extension_name + "  extension!\n";
```

File: scripts/generate_extensions_function.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+functions = {}
+reader = csv.reader(open(os.path.join("..",'extensions.csv')))
+# This skips the first row (i.e., the header) of the CSV file.
+next(reader)
+
+get_func = "select distinct on(function_name) function_name from duckdb_functions();"
+duckdb_path = os.path.join("..",'build', 'release', 'duckdb')
+base_functions = os.popen(f'{duckdb_path} -csv -c "{get_func}" ').read().split("\n")[1:-1]
+
+base_functions = {x for x in base_functions}
```
Comment (Mause) [inline scripts/generate_extensions_function.py:25]:
```suggestion
base_functions = set(base_functions)
```

File: scripts/generate_extensions_function.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    file.write("namespace duckdb { \n\n")
+    file.write("struct ExtensionFunction { \n")
+    file.write("char extension[48]; \n")
+    file.write("char function[48]; \n")
+    file.write("}; \n")
+    file.write("static constexpr ExtensionFunction EXTENSION_FUNCTIONS[] = {\n")
+
+    for function_name in function_map:
+        file.write("{")
+        file.write(f"\"{function_name}\", \"{function_map[function_name]}\"")
```
Comment (Mause) [inline scripts/generate_extensions_function.py:72]:
```suggestion
    for function_name, function_ext in function_map.items():
        file.write("{")
        file.write(f'"{function_name}", "function_ext}"')
```

File: scripts/generate_extensions_function.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+get_func = "select distinct on(function_name) function_name from duckdb_functions();"
+duckdb_path = os.path.join("..",'build', 'release', 'duckdb')
+base_functions = os.popen(f'{duckdb_path} -csv -c "{get_func}" ').read().split("\n")[1:-1]
+
+base_functions = {x for x in base_functions}
+
+function_map = {}
+for row in reader:
+    extension_name = row[0]
```
Comment (Mause) [inline scripts/generate_extensions_function.py:29]:
```suggestion
for extension_name, in reader:
```

File: scripts/generate_extensions_function.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    print("Install/Load " + extension_name)
+    install = f"INSTALL {extension_name};"
+    load = f"LOAD {extension_name};"
+    extension_functions = os.popen(f'{duckdb_path} -csv -c "{install}{load}{get_func}" ').read().split("\n")[1:-1]
+    extension_functions = {x for x in extension_functions}
+    extension_functions = extension_functions.difference(base_functions)
+    for extension_function in extension_functions:
+        function_map[extension_function] = extension_name
```
Comment (Mause) [inline scripts/generate_extensions_function.py:37]:
```suggestion
    function_map.update({
        extension_function: extension_name
        for extension_function in (set(extension_functions) - base_functions)
    })
```

**Review by Mause — CHANGES_REQUESTED** [review]

File: src/main/client_context.cpp [hunk 6]
```diff
@@ -60,6 +61,9 @@ struct ActiveQueryContext {
 ClientContext::ClientContext(shared_ptr<DatabaseInstance> database)
     : db(move(database)), transaction(db->GetTransactionManager(), *this), interrupted(false),
       client_data(make_unique<ClientData>(*this)) {
+	for (auto &functions : EXTENSION_FUNCTIONS) {
```
Comment (Mytherin) [inline src/main/client_context.cpp:64]: This seems quite wasteful, especially as the list of functions will grow and lookups are rarely needed. Can we instead do a lookup in the list of extension names when this is actually required instead, similar to how the default functions and default views work? Perhaps the names can be stored in alphabetical order so a binary search can be done?

File: scripts/generate_extensions_function.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    pattern = re.compile("{\"(.*?)\", \"(.*?)\"},")
+    for line in file:
+        if pattern.match(line):
+            split_line = line.split("\"")
+            cur_function_map[split_line[1]] = split_line[3]
+    assert cur_function_map == function_map
+else:
+    # Generate Header
+    file = open(os.path.join("..","src","include","extension_functions.hpp"),'w')
+    file.write("//===----------------------------------------------------------------------===//\n")
```
Comment (Mytherin) [inline scripts/generate_extensions_function.py:51]: Can this be a multi-line string instead, e.g.:

```py

"""
...
"""

```

That should make the code here a lot more readable.

File: extensions.csv [hunk 8]
```diff
@@ -0,0 +1,13 @@
+name,url,commit,build_on_windows
+excel,https://github.com/duckdb/duckdb,,true
```
Comment (Mytherin) [inline extensions.csv:2]: Perhaps just leave the URL empty for the non-external extensions?

File: src/catalog/catalog.cpp [hunk 9]
```diff
@@ -243,7 +243,14 @@ CatalogException Catalog::CreateMissingEntryException(ClientContext &context, co
 		}
 	});
 	auto unseen_entry = SimilarEntryInSchemas(context, entry_name, type, unseen_schemas);
-
+	if (context.extensions_functions.find(entry_name) != context.extensions_functions.end()) {
+		auto extension_name = context.extensions_functions.find(entry_name)->second;
+		string exception_message = "Function with name " + entry_name +
```
Comment (Mytherin) [inline src/catalog/catalog.cpp:248]: Perhaps use formatting in the exception constructor instead, e.g.:

```return CatalogException("Function with name %s..", entry_name);```

File: src/catalog/catalog.cpp [hunk 10]
```diff
@@ -243,7 +243,14 @@ CatalogException Catalog::CreateMissingEntryException(ClientContext &context, co
 		}
 	});
 	auto unseen_entry = SimilarEntryInSchemas(context, entry_name, type, unseen_schemas);
-
+	if (context.extensions_functions.find(entry_name) != context.extensions_functions.end()) {
```
Comment (Mytherin) [inline src/catalog/catalog.cpp:246]: Can we add a test that triggers this error message?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good, some more comments:

**Pull request merged**